{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week04 data creation\n",
    "\n",
    "Make directories and change to the directory:\n",
    "\n",
    "    mkdir -p ~/genomes/mm10/gencode/m8\n",
    "    cd ~/genomes/mm10/gencode/m8\n",
    "\n",
    "Obtain the data:\n",
    "\n",
    "    wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M8/gencode.vM8.basic.annotation.gtf.gz\n",
    "    wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M8/gencode.vM8.pc_transcripts.fa.gz\n",
    "    wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M8/GRCm38.primary_assembly.genome.fa.gz\n",
    "\n",
    "## Data for STAR and `featureCounts`\n",
    "\n",
    "Get just chr11 from the basic annotation:\n",
    "\n",
    "    grep chr11 gencode.vM8.basic.annotation.gtf > gencode.vM8.basic.annotation.chr11.gtf\n",
    "\n",
    "Get just chr11 from the genome fasta file:\n",
    "\n",
    "    samtools faidx GRCm38.primary_assembly.genome.fa \n",
    "    samtools faidx GRCm38.primary_assembly.genome.fa chr11 > GRCm38.primary_assembly.chr11.fa\n",
    "\n",
    "### STAR: Generate genome\n",
    "\n",
    "#### Chr11 subset\n",
    "\n",
    "Have to make the genome directory first. Use `--runDirPerm ALL_RWX` so created directory can be read by all.\n",
    "\n",
    "    mkdir star_chr11 \n",
    "    STAR \\\n",
    "        --runMode genomeGenerate \\\n",
    "        --runThreadN 4 \\\n",
    "        --runDirPerm All_RWX \\\n",
    "        --genomeFastaFiles GRCm38.primary_assembly.chr11.fa \\\n",
    "        --sjdbGTFfile gencode.vM8.basic.annotation.chr11.gtf \\\n",
    "        --genomeDir star_chr11\n",
    "    \n",
    "Output:\n",
    "\n",
    "    Jan 25 12:30:21 ..... Started STAR run\n",
    "    Jan 25 12:30:21 ... Starting to generate Genome files\n",
    "    Jan 25 12:30:24 ... starting to sort  Suffix Array. This may take a long time...\n",
    "    Jan 25 12:30:25 ... sorting Suffix Array chunks and saving them to disk...\n",
    "    Jan 25 12:31:34 ... loading chunks from disk, packing SA...\n",
    "    Jan 25 12:31:37 ... Finished generating suffix array\n",
    "    Jan 25 12:31:37 ... Generating Suffix Array index\n",
    "    Jan 25 12:32:19 ... Completed Suffix Array index\n",
    "    Jan 25 12:32:19 ..... Processing annotations GTF\n",
    "    Jan 25 12:32:20 ..... Inserting junctions into the genome indices\n",
    "    Jan 25 12:32:39 ... writing Genome to disk ...\n",
    "    Jan 25 12:32:41 ... writing Suffix Array to disk ...\n",
    "    Jan 25 12:32:49 ... writing SAindex to disk\n",
    "    Jan 25 12:33:03 ..... Finished successfully\n",
    "    \n",
    "#### Whole genome\n",
    "\n",
    "    mkdir star\n",
    "    STAR --runMode genomeGenerate --runThreadN 4 --runDirPerm All_RWX --genomeFastaFiles  GRCm38.primary_assembly.genome.fa --sjdbGTFfile gencode.vM8.basic.annotation.gtf --genomeDir star\n",
    "\n",
    "Output:\n",
    "\n",
    "    Jan 25 12:35:22 ..... Started STAR run\n",
    "    Jan 25 12:35:22 ... Starting to generate Genome files\n",
    "    Jan 25 12:36:32 ... starting to sort  Suffix Array. This may take a long time...\n",
    "    Jan 25 12:36:55 ... sorting Suffix Array chunks and saving them to disk...\n",
    "    Jan 25 13:10:36 ... loading chunks from disk, packing SA...\n",
    "    Jan 25 13:16:30 ... Finished generating suffix array\n",
    "    Jan 25 13:16:30 ... Generating Suffix Array index\n",
    "    Jan 25 13:20:25 ... Completed Suffix Array index\n",
    "    Jan 25 13:20:25 ..... Processing annotations GTF\n",
    "    Jan 25 13:20:36 ..... Inserting junctions into the genome indices\n",
    "    Jan 25 13:24:31 ... writing Genome to disk ...\n",
    "    Jan 25 13:25:01 ... writing Suffix Array to disk ...\n",
    "    Jan 25 13:29:56 ... writing SAindex to disk\n",
    "    Jan 25 13:30:13 ..... Finished successfully\n",
    "\n",
    "\n",
    "## Data for `kallisto`\n",
    "\n",
    "Get the gene IDs from chr11:\n",
    "\n",
    "    cut -f 9 gencode.vM8.basic.annotation.chr11.gtf  | awk -F'; ' '{print $1}' | awk -F'\"' '{print $2}' | sort | uniq > gencode.vM8.chr11.gene_ids.txt\n",
    "\n",
    "\n",
    "Use this list of IDs to extract \n",
    "\n",
    "### Create `kallisto index`\n",
    "\n",
    "Command:\n",
    "\n",
    "    kallisto index --index gencode.vM8.pc_transcripts.kallisto gencode.vM8.pc_transcripts.fa \n",
    "\n",
    "Output: \n",
    "\n",
    "    [build] loading fasta file gencode.vM8.pc_transcripts.fa\n",
    "    [build] k-mer length: 31\n",
    "    [build] warning: clipped off poly-A tail (longer than 10)\n",
    "            from 274 target sequences\n",
    "    [build] warning: replaced 1 non-ACGUT characters in the input sequence\n",
    "            with pseudorandom nucleotides\n",
    "    [build] counting k-mers ... done.\n",
    "    [build] building target de Bruijn graph ...  done \n",
    "    [build] creating equivalence classes ...  done\n",
    "    [build] target de Bruijn graph has 285795 contigs and contains 67545152 k-mers \n",
    "\n",
    "\n",
    "## Fastq subset\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
